src/custom-elements.js

```diff
@@ -8,792 +8,47 @@
-
-      // 15:
-      /** @type {CustomElementDefinition} */
-      const definition = {
-        name: name,
-        localName: localName,
-        constructor: constructor,
-        connectedCallback: connectedCallback,
-        disconnectedCallback: disconnectedCallback,
-        attributeChangedCallback: attributeChangedCallback,
-        observedAttributes: observedAttributes,
-      };
-
-      // 16:
-      this._definitions.set(localName, definition);
-      this._constructors.set(constructor, localName);
-
-      // 17, 18, 19:
-      this._upgradeDoc();
-
-      // 20:
-      /** @type {Deferred} **/
-      const deferred = this._whenDefinedMap.get(localName);
-      if (deferred) {
-        deferred.resolve(undefined);
-        this._whenDefinedMap.delete(localName);
-      }
-    }
-
-    /**
-     * Returns the constructor defined for `name`, or `null`.
-     *
-     * @param {string} name
-     * @return {Function|undefined}
-     */
-    get(name) {
-      // https://html.spec.whatwg.org/multipage/scripting.html#custom-elements-api
-      const def = this._definitions.get(name);
-      return def ? def.constructor : undefined;
-    }
-
-    /**
-     * Returns a `Promise` that resolves when a custom element for `name` has
-     * been defined.
-     *
-     * @param {string} name
-     * @return {!Promise}
-     */
-    whenDefined(name) {
-      // https://html.spec.whatwg.org/multipage/scripting.html#dom-customelementsregistry-whendefined
-      const nameError = checkValidCustomElementName(name);
-      if (nameError) return Promise.reject(nameError);
-      if (this._definitions.has(name)) return Promise.resolve();
-
-      /** @type {Deferred} **/
-      let deferred = this._whenDefinedMap.get(name);
-      if (deferred) return deferred.promise;
-
-      let resolve;
-      const promise = new Promise(function(_resolve, _) {
-       resolve = _resolve;
+import CustomElementInternals from './CustomElementInternals';
+import CustomElementRegistry from './CustomElementRegistry';
+import DocumentConstructionObserver from './DocumentConstructionObserver';
+
+import PatchHTMLElement from './Patch/HTMLElement';
+import PatchDocument from './Patch/Document';
+import PatchNode from './Patch/Node';
+import PatchElement from './Patch/Element';
+
+const priorCustomElements = window['customElements'];
+
+if (!priorCustomElements || priorCustomElements['forcePolyfill']) {
+  /** @type {!CustomElementInternals} */
+  const internals = new CustomElementInternals();
+
+  PatchHTMLElement(internals);
+  PatchDocument(internals);
+  PatchNode(internals);
+  PatchElement(internals);
+
+  /** @type {!CustomElementRegistry} */
+  const customElements = new CustomElementRegistry(internals);
+
+  // If `customElements.documentReady` exists and is a Promise, prevent calls to
+  // define from causing upgrades until `documentReady` has resolved.
+  if (priorCustomElements) {
+    const documentReady = priorCustomElements['documentReady'];
```

Comment:
I think we should change the design of this hook to be a function that can return true or a promise. The reason is that this state is dynamic (lazy loading) and we want it to be synchronous so that upgrades happen synchronously when possible.